### PR TITLE
Add a SyntaxVerifier to assert no unknown syntax in a fully parsed tree.

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please test

File: Sources/SwiftSyntax/SyntaxRewriter.swift.gyb [hunk 1]
```diff
@@ -256,3 +256,45 @@ extension Syntax {
...
+    switch self {
+      case .unknownSyntaxFound(let node):
+        return "unknown syntax node for \"\(node)\""
+    }
+  }
+}
+
+public class SyntaxVerifier: SyntaxVisitor {
+
+  var UnknownNodes: [Syntax] = []
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxRewriter.swift.gyb]: Nitpick: Don't capitalise variable names in Swift

File: Sources/SwiftSyntax/SyntaxRewriter.swift.gyb [hunk 2]
```diff
@@ -256,3 +256,45 @@ extension Syntax {
...
+  }
+
+  override public func visitPre(_ node: Syntax) {
+    assert(node.isUnknown)
+    UnknownNodes.append(node)
+  }
+
+  private func verify(_ node: Syntax) throws {
+    node.walk(self)
+    if !UnknownNodes.isEmpty {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxRewriter.swift.gyb]: You could just use optional binding here which get's rid of the force unwrap below

```swift
if let unknownNode = UnknownNodes.first {
  throw SyntaxVerifierError.unknownSyntaxFound(node: unknownNode)
}
```

**Pull request merged**